src/factorio_randovania_mod/factorio_formats.py

```diff
@@ -0,0 +1,147 @@
+import typing
+
+import construct
+
+
+class DictAdapter(construct.Adapter):
+    def _decode(self, obj: construct.ListContainer, context: construct.Container, path: str) -> construct.Container:
+        result: construct.Container = construct.Container()
+        last = {}
+        for i, item in enumerate(obj):
+            key = item.key
+            if key in result:
+                raise construct.ConstructError(f"Key {key} found twice in object", path)
+            last[key] = i
+            result[key] = item.value
+        return result
+
+    def _encode(self, obj: construct.Container, context: construct.Container, path: str) -> construct.ListContainer:
+        return construct.ListContainer(construct.Container(key=type_, value=item) for type_, item in obj.items())
+
+
+class ErrorWithMessage(construct.Construct):
+    def __init__(self, message, error=construct.ExplicitError):
+        super().__init__()
+        self.message = message
+        self.flagbuildnone = True
+        self.error = error
+
+    def _parse(self, stream, context, path):
+        message = construct.evaluate(self.message, context)
+        raise self.error(f"Error field was activated during parsing with error {message}", path=path)
+
+    def _build(self, obj, stream, context, path):
+        message = construct.evaluate(self.message, context)
+        raise self.error(f"Error field was activated during building with error {message}", path=path)
+
+    def _sizeof(self, context, path):
+        raise construct.SizeofError("Error does not have size, because it interrupts parsing and building", path=path)
+
+
+PropertyTreeType = construct.Enum(
```

Comment:
I'd prefer if we used the alternative syntax for `construct.Enum` where we make an `enum.Enum` class and pass it in. that way we can eliminate some of the annoying magic strings hanging around. unless there's some particular reason we want these to be strings?

Reply:
I think I tried and things didn't work out. Can try again

Reply:
Ah it was related to how construct.Enum when using the enum, doesn't give me the enum as parsed result, which caused issues with the Switch below and def `_python_value_to_tree_type`.

Reply:
Did it the closest I can do without having an adapter for `construct.Enum`